net: ethernet: Ensure packet is not touched after consumed by l3

clamattia opened the pull request:
Ensure that all packets are no longer used by ethernet_recv after it was consumed by the l3-handler.

`!net_eth_is_addr_multicast((struct net_eth_addr *)lladdr->addr) && ` is removed. It can never happen.

Stats now reflect ethernet-multicast instead of ip-multicast.

Commit:
net: ethernet: Ensure packet is not touched after consumed by l3

Ensure that all packets are no longer used by ethernet_recv after it was
consumed by the l3-handler.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

jukkar commented:
> Stats now reflect ethernet-multicast instead of ip-multicast.

I don't think the code ever reflected ip-multicast, all the checks in the functions are for eth-multicast addresses.

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -384,20 +376,18 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 	}
 
+out:
+	eth_stats_update_bytes_rx(iface, body_len + hdr_len);
+	eth_stats_update_pkts_rx(iface);
+	if (dst_broadcast) {
+		eth_stats_update_broadcast_rx(iface);
+	} else if (dst_eth_multicast) {
+		eth_stats_update_multicast_rx(iface);
+	}
```
We could add here
```
if (IS_ENABLED(CONFIG_NET_STATISTICS_ETHERNET)) {
	eth_stats_update_bytes_rx(iface, body_len + hdr_len);
	eth_stats_update_pkts_rx(iface);
	if (dst_broadcast) {
		eth_stats_update_broadcast_rx(iface);
	} else if (dst_eth_multicast) {
		eth_stats_update_multicast_rx(iface);
	}
}
```
to avoid checks for `dst_broadcast` or `dst_eth_multicast`. Compiler will probably optimize the calls away but lets make it explicit.

clamattia replied:
This adds a bit noise, I would prefer not to, but will do if you insist.

jukkar replied:
The use of `IS_ENABLED()` is very common in zephyr, it is preferred if the option is not set and we want to compile test the lines of code but want to guarantee that the lines are never there in the binary.

clamattia replied:
Done, so I reverted to the helper function.

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -364,7 +353,10 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 
 		verdict = l3->handler(iface, type, pkt);
-		if (verdict == NET_DROP) {
+		if (verdict == NET_OK) {
+			/* the packet was consumed by the l3-handler */
+			goto out;
```
This now changes the semantics of how the return value is interpreted. So current code allows handler to return either `NET_OK` or `NET_CONTINUE` which are considered ok. If there is no handler, then we go to the unknown type handler. I suppose we could allow the handler to return `NET_CONTINUE` if it does not want to handle the packet, not sure about the use cases in this case. If we go this route, then we need to document this in `net_core.h`. Also current callbacks need to be checked that they work as expected.

clamattia replied:
My understanding is
`NET_OK`, the handler has taken ownership of the packet, everything is fine, handling should not continue
`NET_CONTINUE`, the handler does not take ownership, everything is fine, handling should proceed to the next layer.
`NET_DROP`, the handler does not take ownership, something invalid (for example multicast address not matching protocol) handling should not continue (also not to the next layer).

This is also how I understood the `arp` handler to work for example. I implicitly assumed, that if I have taken ownership of the packet in my handler, that I need to return `NET_OK`, which let to problems. Its also what all in-tree handlers actually do.

jukkar replied:
Ok, it is logical and makes sense here. Currently there is no use case for having multiple ptype handlers but perhaps there is in the future. Please add a commit to this PR that documents this behavior in `net_core.h`, so here https://github.com/zephyrproject-rtos/zephyr/blob/87f11d7efac17c279e7cf19f03f0f0e85173b628/include/zephyr/net/net_core.h#L172

clamattia replied:
Done, please let me know, if the wording is clear.

clamattia replied:
I have assumed it is only possible to have a single handler per ptype. Otherwise it will become complicated, we need to take a `ref` before calling the handler with the packet maybe. Also questions have to be considered like what if one handler wants to drop the packet? Does the other still get it?

jukkar replied:
Yes, these questions are tricky, that is why I considered originally that only one ptype handler can be registered especially as there is no real use cases for having multiple handlers for the same ptype.
But we can tackle this problem in the future. It is still better to document the behavior properly.

jukkar commented:
> > > Stats now reflect ethernet-multicast instead of ip-multicast.
> > 
> > 
> > I don't think the code ever reflected ip-multicast, all the checks in the functions are for eth-multicast addresses.
> 
> The function is a misnomer it actually only checks for ip-multicasts. https://en.wikipedia.org/wiki/Multicast_address#Ethernet :

Miscommunication here, I thought you were talking about IP multicast, but you were talking about Ethernet multicast address for IP. My bad, no issues here.

Commit:
net: core: document the return value of the l3-handler

Give a more detailed explanation of the expectations towards a handler
registered with the NET_L3_REGISTER-facilities.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

clamattia commented:
fixup for typo

clamattia commented:
compliance fixup

clamattia commented:
> Miscommunication here, I thought you were talking about IP multicast, but you were talking about Ethernet multicast address for IP. My bad, no issues here.

Miscommunication also from my side, I did not realize, that there is naturally also a multicast on the level of IP-addresses.

clamattia commented on include/zephyr/net/net_core.h:
```diff
@@ -169,7 +169,14 @@ struct net_l3_register {
 	const struct net_l2 * const l2;
-	/** Handler function for the given protocol type */
+	/** Handler function for the specified protocol type. If the handler
+	 * has taken ownership of the pkt, it must return NET_OK. If it wants to
```
I am referring to these lines of `net_core.c` here:
```C
		if (ret != NET_CONTINUE) {
			if (ret == NET_DROP) {
				NET_DBG("Packet %p discarded by L2", pkt);
				net_stats_update_processing_error(
							net_pkt_iface(pkt));
			}

			return ret;
		}
	}

	net_pkt_set_l2_processed(pkt, true);

	/* L2 has modified the buffer starting point, it is easier
	 * to re-initialize the cursor rather than updating it.
	 */
	net_pkt_cursor_init(pkt);

	if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET_DGRAM)) {
		/* Consecutive call will forward packets to SOCK_DGRAM packet sockets
		 * (after L2 removed header).
		 */
		ret = net_packet_socket_input(pkt, ETH_P_ALL);
		if (ret != NET_CONTINUE) {
			return ret;
		}
	}

	uint8_t family = net_pkt_family(pkt);

	if (IS_ENABLED(CONFIG_NET_IP) && (family == AF_INET || family == AF_INET6 ||
					  family == AF_UNSPEC || family == AF_PACKET)) {
		/* L2 processed, now we can pass IPPROTO_RAW to packet socket:
		 */
		ret = net_packet_socket_input(pkt, IPPROTO_RAW);
		if (ret != NET_CONTINUE) {
			return ret;
		}

		/* IP version and header length. */
		uint8_t vtc_vhl = NET_IPV6_HDR(pkt)->vtc & 0xf0;

		if (IS_ENABLED(CONFIG_NET_IPV6) && vtc_vhl == 0x60) {
			return net_ipv6_input(pkt, is_loopback);
		} else if (IS_ENABLED(CONFIG_NET_IPV4) && vtc_vhl == 0x40) {
			return net_ipv4_input(pkt, is_loopback);
		}

		NET_DBG("Unknown IP family packet (0x%x)", NET_IPV6_HDR(pkt)->vtc & 0xf0);
		net_stats_update_ip_errors_protoerr(net_pkt_iface(pkt));
		net_stats_update_ip_errors_vhlerr(net_pkt_iface(pkt));
		return NET_DROP;
	} else if (IS_ENABLED(CONFIG_NET_SOCKETS_CAN) && family == AF_CAN) {
		return net_canbus_socket_input(pkt);
	}

	NET_DBG("Unknown protocol family packet (0x%x)", family);
```

In my view, these could be also implemented as `NET_L3_REGISTER`, but currently it is not done this way. I was not sure, how far you intended to go with this way of registering the `L3` to the `L2`.

jukkar replied:
I did not want to do this L3 register for IPv4 and IPv6 protocols. This is a can of worms as those protocols are also used with other technologies than Ethernet so would require major changes all over the places. As the IPv4 and IPv6 protocols are pretty much the corner stone of the networking support, we can at least for time being handle them in hard coded way in `net_core.c`

clamattia replied:
Fair, so how should the wording be?
If (in theory) the handler wanted to `continue processing at the next level (e.g. ipv4)` it would have to `return NET_CONTINUE`?
Lets say, I make a custom `L2`, that has their own notion of protocol-type. Would it not be consistent to set the family in the handler and then return `NET_CONTINUE` from the handler so that `net_core` can send it to the appropriate IP-stack?

jukkar replied:
Yes, the handler needs to return `NET_CONTINUE` to allow the stack to process it in further. Hmm, perhaps your wording is ok after all, it is just that the `NET_CONTINUE` has two purpose in the code now. If handler returns `NET_CONTINUE`, then next handler is searched and it can process the packet too. But this value is also used to in net_core to decide if the packet should be processed further. Not sure how to write this in simple way.
The original code did not check `NET_OK` as there could be only one handler. Wondering should we remove the NET_OK check, in which case your comment is very accurate and we do not need to deal with the same ptype used by two handlers? The "multiple handler problem" would go away this way and we could enhance this in the future.

jukkar replied:
Perhaps the last else is not needed as the other branches have gotos, I think the compliance checker will tell that there is unnecessary else here. But we could add the comment telling what happens if NET_CONTINUE is returned.
```
		verdict = l3->handler(iface, type, pkt);
		if (verdict == NET_OK) {
			/* the packet was consumed by the l3-handler */
			goto out;
		} else if (verdict == NET_DROP) {
			NET_DBG("Dropping frame, packet rejected by %s", l3->name);
			goto drop;
		}

		/* The packet will be processed further by IP-stack
		 * when NET_CONTINUE is returned
		 */
		handled = true;
		break;
	}
```

clamattia commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -149,16 +149,16 @@ static inline void ethernet_update_length(struct net_if *iface,
+static void ethernet_update_rx_stats(struct net_if *iface, size_t length,
+				     bool dst_broadcast, bool dst_eth_multicast)
 {
 #if defined(CONFIG_NET_STATISTICS_ETHERNET)
```
Maybe the compiler will no complain here do to the `IS_ENABLED` not being done at preprocessing time. Maybe it would complain, if it was done with `#if` instead. I can go with your suggestion.

clamattia replied:
Sorry, I was confused.
I made a separate PR about this, so this one does not get potentially blocked by the nit: https://github.com/zephyrproject-rtos/zephyr/pull/84928 (note it contains the commits from this PR already).

clamattia commented:
- add additional explanatory comment
- fix rebase mistake
